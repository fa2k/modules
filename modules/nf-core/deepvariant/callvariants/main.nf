
process DEEPVARIANT_CALLVARIANTS {
    tag "$meta.id"
    label 'process_high'

    //Conda is not supported at the moment
    container "nf-core/deepvariant:1.5.0"

    input:
    tuple val(meta), val(intervals), val(make_example_tfrecord_filename), path(make_examples_tfrecords)
    val model_type

    output:
    tuple val(meta), val(intervals), path("call_variants_output.tfrecord.gz"),  emit: call_variants_tfrecords
    path "versions.yml"                                                       ,  emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    // Exit if running this module with -profile conda / -profile mamba
    if (workflow.profile.tokenize(',').intersect(['conda', 'mamba']).size() >= 1) {
        error "DEEPVARIANT module does not support Conda. Please use Docker / Singularity / Podman instead."
    }
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    def model_type_clean = model_type.replaceAll("[^A-Za-z0-9_-]", "")

    """
    /opt/deepvariant/bin/call_variants \\
        --outfile call_variants_output.tfrecord.gz \\
        --examples ${make_example_tfrecord_filename} \\
        --checkpoint "/opt/models/${model_type_clean}/model.ckpt"

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        deepvariant_callvariants: \$(echo \$(/opt/deepvariant/bin/run_deepvariant --version) | sed 's/^.*version //; s/ .*\$//' )
    END_VERSIONS
    """
}
