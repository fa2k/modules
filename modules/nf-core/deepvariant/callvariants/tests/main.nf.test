nextflow_process {

    name "Test Process DEEPVARIANT_CALLVARIANTS"
    script "../main.nf"
    process "DEEPVARIANT_CALLVARIANTS"

    tag "deepvariant/callvariants"
    tag "deepvariant"
    tag "modules"
    tag "modules_nfcore"

    test("homo_sapiens - wgs") {
        setup {
            run("DEEPVARIANT_MAKEEXAMPLES") {
                script "../../makeexamples/main.nf"
                process {
                    """
                    input[0] = [
                        [ id:'test', single_end:false ], // meta map
                        file(params.modules_testdata_base_path + '/genomics/homo_sapiens/illumina/bam/test.paired_end.sorted.bam', checkIfExists: true),
                        file(params.modules_testdata_base_path + '/genomics/homo_sapiens/illumina/bam/test.paired_end.sorted.bam.bai', checkIfExists: true),
                        []
                    ]
                    input[1] = [
                        [ id:'genome'],
                        file(params.modules_testdata_base_path + '/genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true)
                    ]
                    input[2] = [
                        [ id:'genome'],
                        file(params.modules_testdata_base_path + '/genomics/homo_sapiens/genome/genome.fasta.fai', checkIfExists: true)
                    ]
                    input[3] = [
                        [],[]
                    ]
                    """
                }
            }
        }
        when {
            process {
                """
                input[0] = DEEPVARIANT_MAKEEXAMPLES.out.examples
                input[1] = [
                    "wgs" // select model
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert process.out.call_variants_tfrecords.get(0).get(1) == [ id:'test', single_end:false ] },
                // The make_examples_id and the tfrecord binary representation are not stable, though the variant calls are.
                // We can't snapshot the calls tfrecord. See deepvariant subworkflow test and postprocessvariants test.
                { assert process.out.call_variants_tfrecords.get(0).get(2) != null },
                { assert snapshot(process.out.versions).match("versions") },
            )
        }
    }
}
