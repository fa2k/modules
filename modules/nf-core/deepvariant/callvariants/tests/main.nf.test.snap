{
    "versions": {
        "content": [
            [
                "versions.yml:md5,5ff99ffba1e56e4e919d3dfc2d0f3cbb"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-08-09T11:37:27.66602931"
    },
    "homo_sapiens - wgs - stub": {
        "content": [
            {
                "0": [
                    [
                        "dummy_unique_id",
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.call.tfrecord.gz",
                        "test.call-00000-of-00001.tfrecord.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "1": [
                    "versions.yml:md5,5ff99ffba1e56e4e919d3dfc2d0f3cbb"
                ],
                "call_variants_tfrecords": [
                    [
                        "dummy_unique_id",
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.call.tfrecord.gz",
                        "test.call-00000-of-00001.tfrecord.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,5ff99ffba1e56e4e919d3dfc2d0f3cbb"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-08-09T11:37:35.048004009"
    }
}